sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteGroupingAnalyticsWithoutKeys.scala

```diff
@@ -0,0 +1,142 @@
+ *       output = ['k1, 'k2, 'cnt],
+ *       child = Expand(
+ *         projections = [('k1, 'k2, 'v, 0)],
+ *         output = ['k1, 'k2, 'v, 'spark_grouping_id]),
+ *         child = LocalTableScan [...])),
+ *     Aggregate(
+ *       groupingExprs = [],
+ *       aggregateExprs = [null, null, COUNT('v)],
+ *       output = ['k1, 'k2, 'cnt],
+ *       child = Project(
+ *         projectList = ['v],
+ *         child = LocalTableScan [...]))]
+ * }}}
+ */
+object RewriteGroupingAnalyticsWithoutKeys extends Rule[LogicalPlan] {
+
+  private val virtualGroupingColumnNames =
+    Seq(VirtualColumn.groupingIdName, VirtualColumn.groupingPosName)
+
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.transformUp {
+    case agg @ Aggregate(groupingExprs, aggExprs, expand @ Expand(projs, childOutput, child)) =>
+      val groupByAttrMap = AttributeSet.fromAttributeSets(groupingExprs.map {
+        case ar: AttributeReference if virtualGroupingColumnNames.contains(ar.name) =>
+          AttributeSet.empty
+        case e =>
+          e.references
+      })
+      val (groupsWithoutKeys, groupsWithKeys) = {
+        val groupByIndices = childOutput.zipWithIndex.filter { case (a, _) =>
+          groupByAttrMap.contains(a)
+        }
+        projs.partition { groupingSet =>
+          groupByIndices.map(_._2).map(groupingSet).forall {
+            // Checks if all entries in `groupingSet` have NULL
+            case Literal(null, _) => true
+            case _ => false
+          }
+        }
+      }
+      if (groupsWithoutKeys.nonEmpty) {
+        val aggsWithtoutKeys = {
+          val newAggExprs = aggExprs.map {
+            case ne if ne.collectFirst { case ae: AggregateExpression => ae }.nonEmpty => ne
+            // There are the two parts that we need to rewrite in non-aggregate exprs;
+            //  1. Replace group-by attributes with literals with NULL
+            //  2. Fill in a grouping ID attribute for an empty grouping set
+            case ne => ne.transform {
+              case a: Attribute if groupByAttrMap.contains(a) =>
+                Alias(Literal.create(null, ne.dataType), a.name)(exprId = a.exprId)
+              case a: Attribute if a.name == VirtualColumn.groupingIdName =>
+                val gidIdx = childOutput.indexWhere(_.name == VirtualColumn.groupingIdName)
+                Alias(groupsWithoutKeys.head(gidIdx), a.name)(exprId = a.exprId)
+            }
+          }
+
+          // Prunes unnecessary references in child output
+          val newChildOutput = {
+            val aggRefMap = AttributeSet.fromAttributeSets(newAggExprs.flatMap { _.collect {
+              case ae: AggregateExpression => ae.references
+            }})
+            childOutput.filter(aggRefMap.contains)
+          }
+
+          val agg = Aggregate(
+            groupingExpressions = Nil,
+            aggregateExpressions = newAggExprs.map(_.asInstanceOf[NamedExpression]),
+            child = Project(newChildOutput, child))
```

Comment:
One possible side-effect of re-run `child` is, if there is some indeterminate expressions, we may get different projection output.

Reply:
Ur, I see... Good suggestion..., I need more time to reconsider the current approach.